### ISSUE TITLE
CloudFlare SSL failed to do Handshake- (Android 6)<= have issues => (Android 5) version only. 

### ISSUE THREAD

**Issue by jkpatel1463** [android-async-http/android-async-http#1286]
Hello There

Please take a look at error log. It throws error in Android Lolipop and Android Marshmellow. Its works nice with Android Nogut(7) & Android Kitkat.

Error Log:
```
W/System.err: javax.net.ssl.SSLHandshakeException: Handshake failed
  W/System.err:     at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:392)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:535)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:403)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:472)
  W/System.err:     at cz.msebera.android.httpclient.conn.scheme.SchemeSocketFactoryAdaptor.connectSocket(SchemeSocketFactoryAdaptor.java:65)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:177)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:145)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:131)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:611)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:446)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:860)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.makeRequest(AsyncHttpRequest.java:146)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.makeRequestWithRetries(AsyncHttpRequest.java:177)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.run(AsyncHttpRequest.java:106)
  W/System.err:     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
  W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
  W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
  W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
  W/System.err:     at java.lang.Thread.run(Thread.java:818)
  W/System.err: Caused by: javax.net.ssl.SSLProtocolException: SSL handshake aborted: ssl=0xaf228400: Failure in SSL library, usually a protocol error
  W/System.err: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure (external/openssl/ssl/s23_clnt.c:770 0xad9086bd:0x00000000)
  W/System.err:     at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
  W/System.err:     at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:320)
  W/System.err: 	... 19 more
  W/System.err: javax.net.ssl.SSLHandshakeException: Handshake failed
  W/System.err:     at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:392)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:535)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:403)
  W/System.err:     at cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:472)
  W/System.err:     at cz.msebera.android.httpclient.conn.scheme.SchemeSocketFactoryAdaptor.connectSocket(SchemeSocketFactoryAdaptor.java:65)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:177)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:145)
  W/System.err:     at cz.msebera.android.httpclient.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:131)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:611)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:446)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:860)
  W/System.err:     at cz.msebera.android.httpclient.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.makeRequest(AsyncHttpRequest.java:146)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.makeRequestWithRetries(AsyncHttpRequest.java:177)
  W/System.err:     at com.loopj.android.http.AsyncHttpRequest.run(AsyncHttpRequest.java:106)
  W/System.err:     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
  W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
  W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
  W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
  W/System.err:     at java.lang.Thread.run(Thread.java:818)
  W/System.err: Caused by: javax.net.ssl.SSLProtocolException: SSL handshake aborted: ssl=0xaf228400: Failure in SSL library, usually a protocol error
  W/System.err: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure (external/openssl/ssl/s23_clnt.c:770 0xad9086bd:0x00000000)
  W/System.err:     at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
  W/System.err:     at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:320)
  W/System.err: 	... 19 more
  W/System.err: java.lang.NullPointerException: Attempt to get length of null array
  W/System.err:     at java.lang.String.<init>(String.java:206)
  W/System.err:     at com.bitrix.mazebits.apputils.AsyncResponseHandler.onFailure(AsyncResponseHandler.java:70)
  W/System.err:     at com.loopj.android.http.AsyncHttpResponseHandler.handleMessage(AsyncHttpResponseHandler.java:359)
  W/System.err:     at com.loopj.android.http.AsyncHttpResponseHandler$ResponderHandler.handleMessage(AsyncHttpResponseHandler.java:510)
  W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
  W/System.err:     at android.os.Looper.loop(Looper.java:145)
  W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6134)
  W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
  W/System.err:     at java.lang.reflect.Method.invoke(Method.java:372)
  W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
  W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

Source code which generate client for everyrequest:

```
public static AsyncHttpClient newRequest(Context activity) {
        AsyncHttpClient client = new AsyncHttpClient();
        client.addHeader("User-Agent", "android ");
        client.addHeader("Accept", "application/json");
        client.setTimeout(AppConstants.TIMEOUT);
        return client;
    }
```

I am facing issue with some device only(Android Os greater than Lollipop(5) and less than Marshmallow(6)) and with Cloudflare only.

**Comment by nsmith1024** [comment]
Did you solve this problem? i have the same problem. I have android app that connects thru SSL socket to server, it stop working when i started hosting server through cloud flare

**Comment by jkpatel1463** [comment]
Hello @nsmith1024 

After spending time of a day, i figure out that, there have no solution for this lib, so i have replaced it OkHttp & Fast Android Networking Library.

OkHttp: http://square.github.io/okhttp/
Fast Android Networking Library: https://github.com/amitshekhariitbhu/Fast-Android-Networking

**Comment by smarek** [comment]
Thanks, this will be solved in #1029, where I'll concentrace issues about using new underlying HttpClient library and cross-support of SSL/TLS versions vs. Server supported versions
